### ISSUE TITLE
TBB deprecated atomics

### ISSUE THREAD

**Issue by frispete** [AcademySoftwareFoundation/openvdb#887]
Just a heads up, TBB complains:

```
[  238s] cd /home/abuild/rpmbuild/BUILD/blender-2.91.0/build/intern/cycles/kernel/shaders && /usr/bin/oslc -q -O2 -I"/home/abuild/rpmbuild/BUILD/blender-2.91.0/intern/cycles/kernel/shaders" -I"OSL_SHADER_DIR-NOTFOUND" -o /home/abuild/rpmbuild/BUILD/blender-2.91.0/build/intern/cycles/kernel/shaders/node_separate_rgb.oso /home/abuild/rpmbuild/BUILD/blender-2.91.0/intern/cycles/kernel/shaders/node_separate_rgb.osl
[  238s] In file included from /usr/include/openvdb/tree/LeafBuffer.h:10,
[  238s]                  from /usr/include/openvdb/tree/LeafNode.h:11,
[  238s]                  from /usr/include/openvdb/tree/Tree.h:18,
[  238s]                  from /usr/include/openvdb/Grid.h:12,
[  238s]                  from /usr/include/openvdb/openvdb.h:12,
[  238s]                  from /home/abuild/rpmbuild/BUILD/blender-2.91.0/intern/openvdb/intern/openvdb_transform.h:23,
[  238s]                  from /home/abuild/rpmbuild/BUILD/blender-2.91.0/intern/openvdb/intern/openvdb_transform.cc:20:
[  238s] /usr/include/tbb/atomic.h:21:141: note: '#pragma message: TBB Warning: tbb/atomic.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.'
[  238s]    21 | #pragma message("TBB Warning: tbb/atomic.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.")
```

openvdp 7.1.0
TBB 2020.3

**Comment by Idclip** [comment]
Thanks @frispete, we're aware of this and are working on a fix! It will involve switching all deprecated constructs to the standard. Some more info here: https://jira.aswf.io/browse/OVDB-139

**Comment by Idclip** [comment]
linking https://github.com/AcademySoftwareFoundation/openvdb/issues/932

**Comment by BenFrantzDale** [comment]
Looks like tbb/mutex.h is also going away(?) it’s used in Metadata.cc and can easily be replaced by std::mutex I think. Should I make another issue for that or keep it as part of this?

**Comment by Idclip** [comment]
@BenFrantzDale it can be kept as part of this - I've pinned #932 which will essentially solve all these issues (as intel removed support for all deprecated constructs in TBB 2021)

**Comment by BenFrantzDale** [comment]
@Idclip Do you have a sense how involved it is to do the atomics? It looks like they only show up in a handful of places. I'm itching to upgrade my tbb version and may be motivated to make these changes...

**Comment by Idclip** [comment]
I haven't looked too closely at this yet but [this document](https://software.intel.com/sites/default/files/managed/b2/d2/TBBRevamp.pdf) seems to detail a very trivial replacement of most types. The only thing I saw to be different was that the memory semantics for tbb atomics were worded slightly differently to the standard (https://en.cppreference.com/w/cpp/atomic/memory_order). We don't use them in too many places as you pointed out, but we need to make sure they behave identically.

**Comment by BenFrantzDale** [comment]
^ This looks plausible https://github.com/AcademySoftwareFoundation/openvdb/commit/22ae4e5fac69b32cb5574c2e6aa02e77db8ae2e9

**Comment by BenFrantzDale** [comment]
Is this now fixed?

**Comment by danrbailey** [comment]
@BenFrantzDale - not until this PR is merged (#1006). That's your PR with a few minor changes layered on top.

**Comment by danrbailey** [comment]
I have just merged this PR into the master branch. We'll be doing an 8.1 release soon which will include this change.

Closing out this issue.

**Comment by Iktomist** [comment]
Despite the fact that this issue is closed, and has been for over two years, I'm still having problems compiling Blender 3.6.1 because of the TBB references in openVDB. That's from the most recently updated version, too.

**Comment by BenFrantzDale** [comment]
> Despite the fact that this issue is closed, and has been for over two years, I'm still having problems compiling Blender 3.6.1 because of the TBB references in openVDB. That's from the most recently updated version, too.

OpenVDB still depends on TBB, just not the old stuff. I don’t know the exact maximum allowed TBB version for OpenVDB, but I’d guess you just need a recent TBB.
